### ISSUE TITLE
update for phoenix 1.6 and esbuild

### ISSUE THREAD

**Issue by martindemello** [joojscript/phoenix_reactify#1]
Now that phoenix 1.6 is released, could you update this repo to work with the new stack? (I tried to do it myself but was unable to.)

**Comment by joojscript** [comment]
Absolutely, tomorrow the fixes will be here! 

Thanks a lot for the request 😁

**Comment by martindemello** [comment]
wow, thanks, wasn't expecting such a quick response!

**Comment by joojscript** [comment]
As promised, the **[changes are already on master](https://github.com/joojscript/phoenix_reactify/commit/967026e3ee4c346390f76bb6eaeb7dfbf75ebaed)**! And you can now just use the release v0.0.5 and it should works like a charm!

Here are a terminal session I recorded to test the compatibility.

[![asciicast](https://asciinema.org/a/ZvdtjWy2bvqk8Q8mEzCv53iBr.svg)](https://asciinema.org/a/ZvdtjWy2bvqk8Q8mEzCv53iBr)

once again, thanks a lot for the request, and let me know if you have any issue! 😄

**Comment by martindemello** [comment]
thanks, that worked like a charm! if you don't mind, could you explain the architecture a bit more? i was trying (and failing) to get this working with esbuild, but it looks like you still have a webpack config without anything actually calling webpack.

**Comment by joojscript** [comment]
Of course! I hope this will be enough but feel free to ask anything else I possibly forgot here 😄! 

## The library application flow

Basically, it works with 6 "simple" steps, which are the ones in `lib/phoenix_reactify.ex`:

```elixir
opts
    |> verify_prerequisites
    |> verify_if_is_phoenix_project
    |> generate_react_project
    |> add_remount
    |> add_tag_to_index_template
    |> run_last_check
```

This steps follow the way I first get a React app inside my Phoenix project, and I realized I could automate this, to save some time (and also, save everyone from some boiler-plate-tedious-coding time 😆 ). So, in short, the steps would be:

1. Generate a react project on destination folder (Phoenix Reactify assumes `/assets/js`).
2. Add all of its dependencies (along with Typescript and all of the needed*, if specified).
3. Add [Remout](https://github.com/rstacruz/remount), and so, use it to create your spa-tag, which will be embedded on the pages we want our application to run (remember you can do it multiple times, see [advanced usage](https://github.com/joojscript/phoenix_reactify/blob/master/ADVANCED_USAGE.md)).
4. Then, we add the new generated (by remount) tag to our index template, this is to make the user see the changes when he runs the project.
5. And, finally, it does an `npm install --force`, because in some cases, with some specific versions, I noticed that, the javascript dependencies where not installed, and so, this should be a "final verifier".

*with "all of the needed" you can read some babel configuration, and other changes we should do in order to get typescript being interpreted and working 😉 

## The library structure

So, to start off with the actual structure, I'll guide your attention towards two specific files: `lib/helpers/helper.ex` and `/lib/versions.ex`.

At `/lib/helpers/helper.ex`, I create some convenience methods, but I'll call your attention to the definition of **@descriptor**, and the definition of the method **compatible?**. Both of them are used in order to check if the current version of some resource (a.k.a any of the ones in "helpers" directory) is compatible with the versions on `lib/versions.ex`.

Then, as it is a __using__ macro, with special behavior, and I `use` it on the other helper files, they extend the method `compatible?`, on which I simply define **available?**, to check whether they both exist on the user system and are up to date with `/lib/versions.ex`.

Take a look at the **Npm helper** implementation of `available?`:

```elixir
def available? do
    try do
      {version, _} = System.cmd("npm", ["--version"])
      compatible?(String.trim(version))
    rescue
      _ -> {:error, :not_installed, @descriptor}
    end
  end
```

And basically, this is the **available?** method I call when I'm at 
```elixir
opts
    |> verify_prerequisites
```
stage.

Here, check the definition:

```elixir
defp verify_prerequisites(opts) do
    IO.write("Verifying Prerequisites...")

    tasks = [
      Task.async(fn -> Helpers.Npm.available?() end),
      Task.async(fn -> Helpers.Elixir.available?() end),
      Task.async(fn -> Helpers.Erlang.available?() end),
      Task.async(fn -> Helpers.Phoenix.available?() end)
    ]
...
```

---

So, basically this was a simple explanation of how the library code is structured, I did it on kind of a rush, just to make an abstraction and help everyone to avoid the boring boiler-plate that comes along with the setup 😃 

Quick fun-fact: At first, my intention was to create a mix.task, this way you could install once on your machine, and then, just run whenever you want `mix phx.reactify [...args]`, but mix tasks must be self-contained, and I use here [Jason](https://github.com/michalmuskala/jason) as a Json decoder, so if I want this to be a mix task, I must implement my own Json parser, which would be very though, but it is already in my roadmap 😉

**Comment by joojscript** [comment]
> Of course! I hope this will be enough but feel free to ask anything else I possibly forgot here smile!
> 
> ## The library application flow
> [...]

Also, about the build itself, and the webpack usage, I metioned this under the [v0.0.5 release page](https://github.com/joojscript/phoenix_reactify/releases/tag/v0.0.5), but basically Phoniex kept the same behavior as before, so, what I do to configure webpack is completely override `webpack.config.js`, with the contet that I pre-tested and got working! 😉

**Comment by martindemello** [comment]
thanks for the detailed explanation :) i'm still puzzled by one bit - at what point does webpack get called to bundle everything? the 1.5 `mix phx.new` created a `webpack` watcher in `config/dev.exs`, but the 1.6 generated code does not have it.

**Comment by joojscript** [comment]
At 1.6 was implemented the [esbuild](https://esbuild.github.io/), which is a js bundler (similar to webpack, but most argue that is faster). And so, Valim did [this implementation](https://github.com/phoenixframework/esbuild) of it to run on Elixir, and so, from Phoenix 1.6+, esbuild is called by default, and so, there is no webpack anymore:

`/config/config.exs:`
```elixir
# Configure esbuild (the version is required)
config :esbuild,
  version: "0.12.18",
  default: [
    args: ~w(js/app.js --bundle --target=es2016 --outdir=../priv/static/assets),
    cd: Path.expand("../assets", __DIR__),
    env: %{"NODE_PATH" => Path.expand("../deps", __DIR__)}
  ]
```

**Comment by martindemello** [comment]
@joojscript i think you're misunderstanding my question :) i see that you add react support to the project by dropping in a `webpack.config.js`, which processes the react code and puts it in the right place. in 1.5, i assume the webpack watcher called webpack, and webpack in turn read the `webpack.config.js` file and processed the react code. 

in 1.6, as you say, there is no more webpack, since it has been replaced by esbuild - so i imagined one would need an entirely new config script, that esbuild would call. but it seems like even without webpack you are still doing everything through a webpack config file. that's where i get confused by the whole thing.

**Comment by joojscript** [comment]
Oh, I see...I got your question now, but the answer is a little bit disappointing...I don't know exactly.

But...

I dig into the bundle generated after running reactify, and `mix phx.server`, so all the dependencies were bundled and I could take a look at the big picture.

I did not found any of webpack configs, but something I did find was that _probably_ esbuild looped on node_modules directory that was created inside `assets` directory after npm install:

```javascript
 // node_modules/react/index.js
  var require_react = __commonJS({
    "node_modules/react/index.js"(exports, module) {
      "use strict";
      if (false) {
   [..]

  // node_modules/scheduler/cjs/scheduler.development.js
  var require_scheduler_development = __commonJS({
    "node_modules/scheduler/cjs/scheduler.development.js"(exports) {
      "use strict";
      if (true) {
        (function() {
         [..]

  // node_modules/scheduler/index.js
  var require_scheduler = __commonJS({
    [..]

  // node_modules/scheduler/cjs/scheduler-tracing.development.js
  var require_scheduler_tracing_development = __commonJS({
    "node_modules/scheduler/cjs/scheduler-tracing.development.js"(exports) {
      "use strict";
      if (true) {
      [..]
```

What I noticed was that `node_modules/react/index.js` is there, and there is no other point that it is coming from, if not the assets folder itself.

So, simply put, at least as this evidence leads me, I think that for Phoenix v1.6.0+, the webpack.config.js is simply unecessary.

And such, what I decided to do, was to remove the builded bundle, remove the `webpack.config.js` file, and run `mix phx.server` again.

So it bundled it all again, and started the server, the result was the following:

![image](https://user-images.githubusercontent.com/49156424/136293411-564e1c06-2a32-48d3-b279-444049a38994.png)

So, basically, yeah, I think this is it. On Phoenix v1.6.0+ it is unacessary.

~ I'll probably do an "`if`" statement to keep the library from creating this file on these versions ~

And once again, thank you very much for the question, and the help here! 😄

**Comment by martindemello** [comment]
oh, interesting :) thanks for digging into it, and for the explanation!
